### DISCUSSION TITLE
Dynamic constraints to fix specific axes

### DISCUSSION THREAD

**Discussion by amosr** [dartsim/dart#1761] [Q&A] [answered]
Hi DART community,

I'm currently learning DART and am trying to model a ladder-climbing robot. The robot has multiple end-effectors which can hang from the rungs. Right now, I'm modelling the hanging end-effectors by adding dynamic ball-joint constraints, which fix the end-effector to a particular point on the ladder and allow free rotation. However, these ball-joint constraints don't exactly model the constraints on the real robot: in reality, the end-effector is able to slide along the ladder rung, and can only really rotate around the ladder rung axis. Is there any way to model such a constraint? I haven't been able to find any examples of this sort of thing yet, so any pointers would be appreciated.

If there's no built-in support for these constraints, it looks like it might be possible to create a new constraint type by subclassing DynamicJointConstraint and implementing some sort of fixed-axis constraints that way. Would this be a reasonable approach?

Thanks,
Amos

**Comment by jslee02** [comment, marked as answer]
Sorry this sat unanswered for so long. This may no longer be needed by the original poster, but for anyone else with a similar question:

For DART 7/main, `RevoluteJointConstraint` covers the hinge-like part of this: it keeps the two constraint anchors coincident and aligns the hinge axes, while leaving rotation about that axis free.

It does not model the full "slide along the rung and rotate about the rung axis" behavior, because the built-in revolute dynamic constraint still fixes the anchor position. That behavior is closer to a cylindrical-style runtime constraint. DART does not currently provide that exact built-in `DynamicJointConstraint`.

For DART 6.x, `BallJointConstraint` and `WeldJointConstraint` are available, but `RevoluteJointConstraint` is not part of the 6.x line. We do not currently plan to add the cylindrical-style constraint to DART 6 LTS by default, because it would be a new feature/API rather than a maintenance fix. For 6.x users, the practical options are to model the mechanism with normal skeleton joints when runtime attach/detach is not required, or implement a custom `DynamicJointConstraint`/`ConstraintBase` for this exact behavior.
